pkg/datapath/linux/probes/probes.go

```diff
@@ -201,85 +208,96 @@ func (p *ProbeManager) SystemKernelHz() (int, error) {
-		return fmt.Errorf(
-			"CONFIG_HAVE_EBPF_JIT kernel parameter is required")
-	}
-	// Optional
-	if !config.ConfigCgroupBpf.Enabled() {
-		log.Warning(
-			"CONFIG_CGROUP_BPF optional kernel parameter is not in kernel configuration")
-	}
-	if !config.ConfigLwtunnelBpf.Enabled() {
-		log.Warning(
-			"CONFIG_LWTUNNEL_BPF optional kernel parameter is not in kernel configuration")
+	requiredParams := p.GetRequiredConfig()
+	for param, kernelOption := range requiredParams {
+		if !kernelOption.Enabled && kernelOption.CanBeModule == true {
+			return fmt.Errorf("%s kernel parameter or module is required", param)
+		} else {
+			return fmt.Errorf("%s kernel parameter is required", param)
+		}
 	}
-	if !config.ConfigBpfEvents.Enabled() {
-		log.Warning(
-			"CONFIG_BPF_EVENTS optional kernel parameter is not in kernel configuration")
+	optionalParams := p.GetOptionalConfig()
+	for param, kernelOption := range optionalParams {
+		if !kernelOption.Enabled && kernelOption.CanBeModule == true {
+			log.Warningf("%s optional kernel parameter or module is not in kernel", param)
+		} else {
+			log.Warningf("%s optional kernel parameter or module is not in kernel", param)
+		}
 	}
 	return nil
 }
 
 // GetRequiredConfig performs a check of mandatory kernel configuration options. It
 // returns a map indicating which required kernel paramaters are enabled - and which are not.
 // GetRequiredConfig is being used by CLI "cilium kernel-check".
-func (p *ProbeManager) GetRequiredConfig() map[KernelParam]bool {
+func (p *ProbeManager) GetRequiredConfig() map[KernelParam]kernelOption {
 	config := p.features.SystemConfig
-	kernelParams := make(map[KernelParam]bool)
+	kernelParams := make(map[KernelParam]kernelOption)
 
-	kernelParams["CONFIG_BPF"] = config.ConfigBpf.Enabled()
-	kernelParams["CONFIG_BPF_SYSCALL"] = config.ConfigBpfSyscall.Enabled()
-	kernelParams["CONFIG_NET_SCH_INGRESS"] = config.ConfigNetSchIngress.Enabled()
-	kernelParams["CONFIG_NET_CLS_BPF"] = config.ConfigNetClsBpf.Enabled()
-	kernelParams["CONFIG_NET_CLS_ACT"] = config.ConfigNetClsAct.Enabled()
-	kernelParams["CONFIG_BPF_JIT"] = config.ConfigBpfJit.Enabled()
-	kernelParams["CONFIG_HAVE_EBPF_JIT"] = config.ConfigHaveEbpfJit.Enabled()
+	kernelParams["CONFIG_BPF"] = kernelOption{
+		Enabled:     config.ConfigBpf.Enabled(),
+		Description: "",
+		CanBeModule: false,
+	}
+	kernelParams["CONFIG_BPF_SYSCALL"] = kernelOption{
+		Enabled:     config.ConfigBpfSyscall.Enabled(),
+		Description: "",
+		CanBeModule: false,
+	}
+	kernelParams["CONFIG_NET_SCH_INGRESS"] = kernelOption{
+		Enabled:     config.ConfigNetSchIngress.Enabled(),
+		Description: "",
+		CanBeModule: true,
+	}
+	kernelParams["CONFIG_NET_CLS_BPF"] = kernelOption{
+		Enabled:     config.ConfigNetClsBpf.Enabled(),
+		Description: "",
+		CanBeModule: true,
+	}
+	kernelParams["CONFIG_NET_CLS_ACT"] = kernelOption{
+		Enabled:     config.ConfigNetClsAct.Enabled(),
+		Description: "",
+		CanBeModule: false,
+	}
```

Comment:
Can't this be a module? Do we need it anyway? I would think we're attaching only to classifiers, not to actions? /cc @joestringer

Reply:
> Can't this be a module? Do we need it anyway? I would think we're attaching only to classifiers, not to actions? /cc @joestringer

@qmonnet hm, if we don't need it anyway, we then might want to remove it alltogheter - i. e. remove it from `SystemConfig` and its dependencies?

Reply:
This is the CLS_ACT, ie classifier -> direct action feature. We are using it:

```
$ git grep clsact
Documentation/bpf.rst:    # tc qdisc add dev em1 clsact
Documentation/bpf.rst:    # tc qdisc add dev em1 clsact
Documentation/bpf.rst:  The first step is to set up a ``clsact`` qdisc (Linux queueing discipline). ``clsact``
Documentation/bpf.rst:  the ``bpf`` classifier. The ``clsact`` qdisc provides two special hooks called
Documentation/bpf.rst:  The ``clsact`` qdisc is processed lockless from ``ingress`` and ``egress``
Documentation/bpf.rst:  For removing the entire ``clsact`` qdisc from the netdevice, which implicitly also
Documentation/bpf.rst:    # tc qdisc del dev em1 clsact
Documentation/bpf.rst:    # tc qdisc add dev em1 clsact
Documentation/bpf.rst:    # tc qdisc add dev em1 clsact
Documentation/bpf.rst:    # tc qdisc add dev em1 clsact
Documentation/bpf.rst:attached to is managed by a pseudo qdisc called ``sch_clsact``. This is a
Documentation/bpf.rst:``sch_clsact`` qdisc's egress hook comes at a much earlier point however which
Documentation/bpf.rst:egress qdiscs. Thus for cases like ``sch_htb`` the ``sch_clsact`` qdisc could
Documentation/bpf.rst:Offloaded tc BPF programs are supported for the case of ``sch_clsact`` in
Documentation/bpf.rst:  The ``sch_clsact``'s egress hook which is called ``sch_handle_egress()``
Documentation/bpf.rst:  qdisc such as ``sch_htb``. This type of interaction of ``sch_clsact``
Documentation/bpf.rst:  allows to reduce the lock contention on transmission since ``sch_clsact``'s
bpf/include/linux/bpf.h: *              Retrieving this identifier works with the clsact TC egress hook
bpf/include/linux/bpf.h: *              clsact TC egress hook, this has the advantage that, internally,
bpf/init.sh:    tc qdisc replace dev $DEV clsact || true
bpf/init.sh:                    tc qdisc del dev $DEV clsact 2> /dev/null || true
pkg/datapath/connector/ipvlan.go:                       QdiscType:  "clsact",
pkg/datapath/connector/ipvlan.go:                       return fmt.Errorf("failed to create clsact qdisc on %q: %s", dstIfName, err)
pkg/datapath/loader/netlink.go:         QdiscType:  "clsact",
test/bpf/verifier-test.sh:              $TC qdisc replace dev ${DEV} clsact
```

Reply:
Of course. Sorry, silly question, I got confused somehow -_-.